### PR TITLE
Add 5 minutes to initial usage reporting delay.

### LINKED ISSUE #1194: Increase initial usage reporting delay to 5 minutes
Will prevent spam in case of cyclic restart.

### PR DESCRIPTION
**Pull request by fasaxc**
Prevents spamming server under cyclic-restart scenarios.

Testing performed:

- Add UT for delay calculation.
- Manual test that sensible delay values printed to log.

Fixes #1194

### CODE REVIEW

File: go/felix/usagerep/usagerep_test.go [hunk 1]
```diff
@@ -60,4 +61,27 @@ var _ = Describe("Usagerep", func() {
 		Expect(q.Get("guid")).To(Equal("baddecaf"))
 		Expect(q.Get("cluster_type")).To(Equal("unknown"))
 	})
+	It("should delay at least 5 minutes", func() {
+		Expect(calculateInitialDelay(0)).To(BeNumerically(">=", 5*time.Minute))
+		Expect(calculateInitialDelay(1)).To(BeNumerically(">=", 5*time.Minute))
+		Expect(calculateInitialDelay(1000)).To(BeNumerically(">=", 5*time.Minute))
+	})
+	It("should have a random component", func() {
+		Expect(calculateInitialDelay(1000000)).ToNot(Equal(calculateInitialDelay(1000000)))
```
Comment (neiljerram) [inline go/felix/usagerep/usagerep_test.go]: This test - with 1 million hosts - seems unrealistic.  Is the number so large because you're trying to reduce the likelihood of a possible collision?  If so, I think a better test would be to call calculateInitialDelay(50) 10 times, and expect that at least one of the values is different.

File: go/felix/usagerep/usagerep_suite_test.go [hunk 2]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package usagerep_test
+package usagerep
```
Comment (neiljerram) [inline go/felix/usagerep/usagerep_suite_test.go]: Why this change?

File: go/felix/usagerep/usagerep.go [hunk 3]
```diff
@@ -54,6 +50,19 @@ func PeriodicallyReportUsage(interval time.Duration, hostname, clusterGUID, clus
 	}
 }
 
+func calculateInitialDelay(numHosts int) time.Duration {
+	// To avoid thundering herd, inject some startup jitter.
+	if numHosts <= 0 {
+		// Intn requires a positive argument so clamp numHosts to be >= 1.
+		numHosts = 1
+	}
+	initialJitter := time.Duration(rand.Intn(numHosts*1000)) * time.Millisecond
```
Comment (neiljerram) [inline go/felix/usagerep/usagerep.go]: Presumably the range of a Go integer is large enough that we don't need to worry about numHosts*1000 overflowing?

Comment (neiljerram) [inline go/felix/usagerep/usagerep.go]: Also I'm wondering if a paren is misplaced here, and it should be:

    initialJitter := time.Duration(rand.Intn(numHosts*1000) * time.Millisecond)

(Because it appears that the other `time.` methods around here don't handle plain integers.)

File: go/felix/usagerep/usagerep_test.go [hunk 4]
```diff
@@ -60,4 +61,27 @@ var _ = Describe("Usagerep", func() {
...
+	})
+	It("should have an average close to expected value", func() {
+		var total time.Duration
+		for i := int64(0); i < 1000000; i++ {
+			total += calculateInitialDelay(60)
+			if i > 100 {
+				average := time.Duration(int64(total) / (i + 1))
+				if average > (5*time.Minute+20*time.Second) &&
```
Comment (neiljerram) [inline go/felix/usagerep/usagerep_test.go]: I think it's worth commenting here that the average jitter is 0.5s per host, and so 30s for 60 hosts - and hence the test for the average being between 20s and 40s.

Comment (neiljerram) [inline go/felix/usagerep/usagerep_test.go]: This is a very nice test, though!

File: go/felix/usagerep/usagerep.go (hunk 3 above)
Comment (fasaxc) [inline go/felix/usagerep/usagerep.go]: Think it's OK, should be good enough for 2m hosts.

File: go/felix/usagerep/usagerep_test.go (hunk 1 above)
Comment (fasaxc) [inline go/felix/usagerep/usagerep_test.go]: Done

File: go/felix/usagerep/usagerep.go (hunk 3 above)
Comment (fasaxc) [inline go/felix/usagerep/usagerep.go]: It's correct: `time.Duration` is a type, not a method.  `time.Duration(...)` is a cast, which is required by go's (rather minimal) promotion rules.  `time.Millisecond` is a `time.Duration` constant.

File: go/felix/usagerep/usagerep_test.go [hunk 5]
```diff
@@ -60,4 +61,42 @@ var _ = Describe("Usagerep", func() {
 		Expect(q.Get("guid")).To(Equal("baddecaf"))
 		Expect(q.Get("cluster_type")).To(Equal("unknown"))
 	})
+	It("should delay at least 5 minutes", func() {
+		Expect(calculateInitialDelay(0)).To(BeNumerically(">=", 5*time.Minute))
+		Expect(calculateInitialDelay(1)).To(BeNumerically(">=", 5*time.Minute))
+		Expect(calculateInitialDelay(1000)).To(BeNumerically(">=", 5*time.Minute))
+	})
+	It("should delay at most 10000 seconds", func() {
```
Comment (neiljerram) [inline go/felix/usagerep/usagerep_test.go]: This should really say '5 minutes + 10000 seconds'.
